js/beaxy.js

```diff
@@ -0,0 +1,414 @@
+'use strict';
+
+//  ---------------------------------------------------------------------------
+
+const Exchange = require ('./base/Exchange');
+
+//  ---------------------------------------------------------------------------
+
+module.exports = class beaxy extends Exchange {
+    describe () {
+        return this.deepExtend (super.describe (), {
+            'id': 'beaxy',
+            'name': 'Beaxy',
+            'countries': [ 'US' ],
+            'rateLimit': 500,
+            'userAgent': this.userAgents['chrome'],
+            'has': {
+                'CORS': false,
+                'fetchMarkets': true,
+                'fetchCurrencies': true,
+                'fetchTickers': true,
+                'fetchTrades': true,
+                'fetchOrderBook': true,
+                'fetchOHLCV': true,
+            },
+            'timeframes': {
+                '1m': 'MINUTE',
+                '5m': 'MINUTE5',
+                '15m': 'MINUTE15',
+                '30m': 'MINUTE30',
+                '1h': 'HOUR',
+                '4h': 'HOUR4',
+                '8h': 'HOUR8',
+                '1d': 'DAY',
+                '1w': 'WEEK',
+            },
+            'urls': {
+                'api': {
+                    'public': 'https://services.beaxy.com/api/v2',
+                },
+                'www': 'https://beaxy.com',
+                'doc': 'https://beaxyapiv2.docs.apiary.io',
+            },
+            'api': {
+                'public': {
+                    'get': [
+                        'symbols',
+                        'currencies',
+                        'symbols/rates',
+                        'symbols/{market}/rate',
+                        'symbols/{market}/trades',
+                        'symbols/{market}/chart',
+                        'symbols/{market}/book',
+                    ],
+                },
```

Comment:
Hi @PurpleTheBest !

Thanks for the PR!

However, it is incomplete at this point. Namely, we will need to add the private endpoints and the corresponding unified methods, as outlined here: https://beaxyapiv2trading.docs.apiary.io/#/reference

Reply:
Hi @kroitor 

It's a bit challenging to integrate trading API tbh. Lemme describe it in two words. The flow starts from opening session by creating session secret (to sign each API request SHA384) using DH and sign (RSA-SHA256) login challenge with private key. Then you should maintain that session by refreshing it every 100 sec if it's inactive. So by providing API key and private key to exchange object force me to maintain all that stuff inside ccxt and it's kinda breaks all rules described in manual. Is that OK if session maintenance goes completely to user side? In this case you can pass session Id/secret as a credentials (or maybe as an options) and everything hopefully will be fine. 

PS: Thanks for comments and remarks. I will fix them ASAP.

Reply:
@PurpleTheBest Hey!

> The flow starts from opening session by creating session secret (to sign each API request SHA384) using DH and sign (RSA-SHA256) login challenge with private key. Then you should maintain that session by refreshing it every 100 sec if it's inactive. 

In general, we add a signIn helper that is used to obtain the access token for a token-based authentication schema you've described (see how it is implemented in the other exchanges throughout the lib, cmd+f or ctrl+f `signIn` in the existing files). After that we add a tracking mechanism in `sign()` that will throw upon expiration, thus signaling the user to redo the `signIn` all upon expiry, and will update the expiration timestamp upon reusing the token.

The second step to it is to add a portable implementation of DH and RSA-SHA256 in JS/Python/PHP (most of that logic is already implemented in the CCXT basecode). When it is done, we have to test the private methods in all languages according to the following: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#multilanguage-support.

However, tbh, I don't really understand why the exchange would choose to make the authentication for the user as difficult as possible – most of the exchanges work fine with HMAC-SHA256 authentication based on API keypairs, and that is secure enough for most of the needs. I'd highly recommend to reconsider the authentication schema on the exchange backend. It is in the best interest of the exchange to make the onboarding for API users as easy and straightforward as possible, that will help gain the necessary traction, and lock-in the userbase. Suppose that some users won't use CCXT – it will be very hard for them to actually implement this type of API authentication and maintain the schema you've described above – that will obviously go against the interest of the exchange in the first place.

Reply:
Yeah I know) But it is separate product by Deltix. Lets say its just temporary solution until we build new API on top of it with ofc much more user friendly interface. Thx for suggestions btw, I missed that JWT auth flow somehow.